### PR TITLE
#3813 Add onChange method in Paper

### PR DESCRIPTION
**Pull request by konekoya**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #3813)
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [x] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxUtils.js [hunk 1]
```diff
@@ -181,12 +181,14 @@ export const enableDragAndDrop = params => {
...
+      const cellKind = cellView.model.get('kind');
+      $('#paper').append(`<div class="flying-paper flying-${cellKind}"></div>`);
+
       const flyingGraph = new joint.dia.Graph();
       new joint.dia.Paper({
-        el: $('#flying-paper'),
-        width: 160,
-        height: 50,
+        el: $('.flying-paper'),
+        width: cellKind === KIND.topic ? 60 : 160,
```
Comment (konekoya) [inline ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxUtils.js]: Change the dragging icon to circle /cc @eechih

File: ohara-manager/client/src/components/Pipeline/Paper/PaperUtils.js [hunk 2]
```diff
@@ -47,6 +48,13 @@ export const getCellData = cellOrView => {
   };
 };
 
+export const getChangeEventType = cell => {
+  if (_.has(cell, 'changed.position')) return 'position';
+  if (_.has(cell, 'changed.target')) return 'connection';
```
Comment (konekoya) [inline ohara-manager/client/src/components/Pipeline/Paper/PaperUtils.js]: @wu87988622 I'll expose more event types later if needed

File: ohara-manager/client/src/components/Pipeline/Paper/Paper.js [hunk 3]
```diff
@@ -482,6 +529,9 @@ const Paper = React.forwardRef((props, ref) => {
 
         return graph.getCells().map(getCellData);
       },
+      load(cellData) {
+        graph.resetCells(cellData);
```
Comment (konekoya) [inline ohara-manager/client/src/components/Pipeline/Paper/Paper.js]: I'll complete this loading method in another task in #3813

File: ohara-manager/client/src/components/Pipeline/Paper/Paper.js [hunk 4]
```diff
@@ -237,24 +252,58 @@ const Paper = React.forwardRef((props, ref) => {
...
-        !_.isEqual(cellRef.current, cell) &&
-        _.get(paperApi, 'state.isReady', false)
-      ) {
-        const data = getCellData(cell);
-        cellRef.current = cell;
-        onCellEventRef.current.onElementAdd(data, paperApi);
-      }
+      if (_.isEqual(cellAddRef.current, cell)) return;
+      if (!_.get(paperApi, 'state.isReady')) return;
+      if (cell.isLink()) return;
```
Comment (konekoya) [inline ohara-manager/client/src/components/Pipeline/Paper/Paper.js]: Adding a new link is not counted a real "add"

File: ohara-manager/client/src/components/Pipeline/Paper/Paper.js [hunk 5]
```diff
@@ -237,24 +252,58 @@ const Paper = React.forwardRef((props, ref) => {
...
 
     // This paperApi should be passed into all event handlers as the second
-    // parameter,  // e.g., onConnect(..., paperApi)
+    // parameter,  // e.g., onConnect(eventObject, paperApi)
     // the only exception is the onRemove as it doesn't any other parameters.
     const paperApi = ref.current;
 
-    // Binding custom handlers
+    // Binding custom event handlers
     graph.on('add', cell => {
```
Comment (konekoya) [inline ohara-manager/client/src/components/Pipeline/Paper/Paper.js]: Add, change and remove together form the custom "onChange" method /cc @wu87988622  :)

File: ohara-manager/client/src/components/Pipeline/Paper/Paper.js [hunk 6]
```diff
@@ -114,6 +125,10 @@ const Paper = React.forwardRef((props, ref) => {
     const graph = graphRef.current;
     const paper = paperRef.current;
 
+    paper.on('cell:pointermove', cell => {
+      cellChangeRef.current = cell;
```
Comment (konekoya) [inline ohara-manager/client/src/components/Pipeline/Paper/Paper.js]: Need to store the ref here so we can tell it needs to be updated or not in the `graph.on('change')` event

File: ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxUtils.js (hunk 1 above)
Comment (eechih) [reply ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxUtils.js]: good.

File: ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxUtils.js [hunk 7]
```diff
@@ -77,8 +78,18 @@ export const createToolboxList = params => {
...
-      this.$box.find('.display-name').text(this.model.get('name'));
-      this.$box.find('.icon').html(this.model.get('icon'));
+
+      const $box = this.$box;
+      const model = this.model;
+
+      $box.find('.display-name').text(model.get('name'));
+      $box.find('.icon').html(model.get('icon'));
+
+      if (model.get('isDisabled') !== undefined) {
```
Comment (konekoya) [inline ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxUtils.js]: Addressed #3871

**Pull request merged**